File: python/pypto/op/matmul.py
#!/usr/bin/env python3
# coding: utf-8
# Copyright (c) 2025 Huawei Technologies Co., Ltd.
# This program is free software, you can redistribute it and/or modify it under the terms and conditions of
# CANN Open Software License Agreement Version 2.0 (the "License").
# Please refer to the License for details. You may not use this file except in compliance with the License.
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
# INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
# See LICENSE in the root of the software repository for the full text of the License.
# -----------------------------------------------------------------------------------------------------------
""" """
import struct

from .. import pypto_impl
from .._op_wrapper import op_wrapper
from ..tensor import Tensor


@op_wrapper
def matmul(
    input,
    mat2,
    out_dtype,
    *,
    a_trans=False,
    b_trans=False,
    c_matrix_nz=False,
    extend_params=None
) -> Tensor:
    """
    Supports two forms of matrix multiplication computation:
    (1) Performs a matrix multiplication of the matrices `input` and `mat2`
    (2) Performs a batch matrix-matrix multiplication of the matrices `input` and `mat2`

    `input` and `mat2` support 2-D or 3-D or 4-D tensors each containing the same number of matrices.
    If `input` is a (n x k) tensor, `mat2` is a (k x m) tensor, output will be a (m x n) tensor.
    If `input` is a (b x n x k) tensor, `mat2` is a (b x k x m) tensor, output will be a (b x m x n) tensor.

    NOTES:
    If `input` of `mat2` is 3-D or 4-D, this function support broadcast.
    For example, if `input` is a (1 x n x k)tensor and `mat2` is a (b x k x m) tensor, the batch dimensions are (1)
    and (b), and the matrix dimensions are (n x k) and (k x m). output will be a (b x m x n) tensor.

    Parameters
    --------
    input : Tensor
        the left matrix to be matrix multiplied.
    mat2 : Tensor
        the right matrix to be matrix multiplied.
    out_dtye : dtype
        the dtype of the output tensor.

    Keyword Arguments
    --------
    a_trans : bool
        whether to transpose the left matrix. Default is False.
    b_trans : bool
        whether to transpose the right matrix. Default is False.
    c_matrix_nz : bool
        whether output matrix is in NZ format. Default is False.
    extend_params: dict
        the extend features of matrix multiplication computation:
        (1) bias: adds a learnable bias to the output.
        keyword arguments: 'bias_tensor': Tensor
        (2) Dequantization: C = DEQF16(ReLu(A @ B))
        keyword arguments: 'scale_tensor': Tensor, 'scale': float, 'relu_type': ReLuType

    Returns
    --------
    Tensor
        A new Tensor containing the matrix multiplied result.

    Raise
    --------
    RuntimeError
        If the dimensions of matrix `input` and `mat2` are not equal and greater than 4-D or less than 2-D.

    Examples
    --------
    # matrix x matrix
    a = pto.tensor((16, 32), pto.DT_BF16, "tensor_a")
    b = pto.tensor((32, 64), pto.DT_BF16, "tensor_b")
    pto.matmul(a, b, pto.DT_BF16)


    # batched matrix multiplication
    a = pto.tensor((2, 16, 32), pto.DT_FP16, "tensor_a")
    b = pto.tensor((2, 32, 16), pto.DT_FP16, "tensor_b")
    pto.matmul(a, b, pto.DT_FP16)


    # batched matrix multiplication with broadcast
    a = pto.tensor((1, 32, 64), pto.DT_FP32, "tensor_a")
    b = pto.tensor((3, 64, 16), pto.DT_FP32, "tensor_b")
    pto.matmul(a, b, pto.DT_DT_FP32)

    # matrix multiplication with bias
    a = pto.tensor((16, 32), ptoDT_FP16, "tensor_a")
    b = pto.tensor((32, 64), pto.DT_FP16, "tensor_b")
    bias = pto.tensor((1, 64), pto.DT_FP16, "tensor_bias")
    extend_params = {'bias_tensor': bias}
    pto.matmul(a, b, pto.DT_BF16, a_trans=False, b_trans=False, c_matrix_nz=False, extend_params=extend_params)

    # matrix multiplication with dequantization
    a = pto.tensor((16, 32), pto.DT_INT8, "tensor_a")
    b = pto.tensor((32, 64), pto.DT_INT8, "tensor_b")
    extend_params = {'scale': 0.2}
    pto.matmul(a, b, pto.DT_BF16, a_trans=False, b_trans=False, c_matrix_nz=False, extend_params=extend_params)

    """
    input_dim = input.Dim()
    mat2_dim = mat2.Dim()
    if input_dim == mat2_dim == 2:
        if (extend_params is None) or (not extend_params):
            return pypto_impl.Matmul(
                out_dtype, input, mat2, a_trans, b_trans, c_matrix_nz
            )
        else:
            extend_params = pypto_impl.MatmulExtendParam(
                **convert_matmul_extend_params(extend_params)
            )
            return pypto_impl.Matmul(
                out_dtype, input, mat2, a_trans, b_trans, c_matrix_nz, extend_params
            )
    elif (input_dim == mat2_dim == 3) or (input_dim == mat2_dim == 4):
        return pypto_impl.BatchMatmul(
            out_dtype, input, mat2, a_trans, b_trans, c_matrix_nz
        )
    else:
        raise RuntimeError(
            "input dim and mat dim must equals, which only support 2-D/3-D/4-D currently"
        )


def convert_matmul_extend_params(extend_params) -> dict:
    extend_params.setdefault('bias_tensor', pypto_impl.Tensor())
    extend_params.setdefault('scale_tensor', pypto_impl.Tensor())
    extend_params.setdefault('relu_type', pypto_impl.ReLuType.NoReLu)
    extend_params.setdefault('scale', 0.0)
    return extend_params

File: framework/src/interface/ir/op/block_ops/matmul.cpp
/*
 * Copyright (c) 2026 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 * -----------------------------------------------------------------------------------------------------------
 */

#include <any>
#include <memory>
#include <string>
#include <vector>

#include "core/logging.h"
#include "ir/expr.h"
#include "ir/kind_traits.h"
#include "ir/op_registry.h"
#include "ir/scalar_expr.h"
#include "ir/type.h"
#include "ir/type_inference.h"

namespace pypto {
namespace ir {

TypePtr DeduceBlockMatMulType(const std::vector<ExprPtr> &args,
    const std::vector<std::pair<std::string, std::any>> &kwargs, const std::string &opName) {
    (void)kwargs;
    CHECK(args.size() == 2) << "The operator " << opName << " requires exactly 2 arguments, but got " << args.size();

    // Both arguments must be TileType
    auto lhsType = As<TileType>(args[0]->GetType());
    auto rhsType = As<TileType>(args[1]->GetType());

    CHECK(lhsType) << "The operator " << opName << " requires first argument to be a TileType, but got "
                   << args[0]->GetType()->TypeName();
    CHECK(rhsType) << "The operator " << opName << " requires second argument to be a TileType, but got "
                   << args[1]->GetType()->TypeName();

    // Extract shapes
    const auto &lhsShape = lhsType->shape_;
    const auto &rhsShape = rhsType->shape_;

    // For block matmul, we require 2D tiles
    CHECK(lhsShape.size() == 2) << "The operator " << opName << " requires lhs to be 2D, but got " << lhsShape.size()
                                << " dimensions";
    CHECK(rhsShape.size() == 2) << "The operator " << opName << " requires rhs to be 2D, but got " << rhsShape.size()
                                << " dimensions";

    // Matrix multiplication: [M, K] @ [K, N] -> [M, N]
    // We need to verify that K dimensions match
    // Note: In PTO ISA, we see [M, K] @ [K, N] -> [M, N]

    ExprPtr mDim = lhsShape[0];
    ExprPtr kDimLhs = lhsShape[1];
    ExprPtr kDimRhs = rhsShape[0];
    ExprPtr nDim = rhsShape[1];

    // Try to verify K dimensions match if they are constant
    auto kLhsConst = As<ConstInt>(kDimLhs);
    auto kRhsConst = As<ConstInt>(kDimRhs);

    if (kLhsConst && kRhsConst) {
        CHECK(kLhsConst->value_ == kRhsConst->value_)
            << "The operator " << opName << " requires matching inner dimensions, but got lhs K=" << kLhsConst->value_
            << " and rhs K=" << kRhsConst->value_;
    }

    // Promote data types
    auto resultDtype = PromoteDataTypes(lhsType->dtype_, rhsType->dtype_);
    CHECK(resultDtype) << "The operator " << opName << " requires compatible data types, but got "
                       << lhsType->dtype_.ToString() << " and " << rhsType->dtype_.ToString();

    // Output shape is [M, N]
    std::vector<ExprPtr> outputShape = {mDim, nDim};

    return std::make_shared<TileType>(outputShape, *resultDtype);
}

TypePtr DeduceBlockMatMulAccType(const std::vector<ExprPtr> &args,
    const std::vector<std::pair<std::string, std::any>> &kwargs, const std::string &opName) {
    (void)kwargs;
    CHECK(args.size() == 3) << "The operator " << opName << " requires exactly 3 arguments, but got " << args.size();

    // All arguments must be TileType
    auto accType = As<TileType>(args[0]->GetType());
    auto lhsType = As<TileType>(args[1]->GetType());
    auto rhsType = As<TileType>(args[2]->GetType());

    CHECK(accType) << "The operator " << opName << " requires first argument (acc) to be a TileType, but got "
                   << args[0]->GetType()->TypeName();
    CHECK(lhsType) << "The operator " << opName << " requires second argument (lhs) to be a TileType, but got "
                   << args[1]->GetType()->TypeName();
    CHECK(rhsType) << "The operator " << opName << " requires third argument (rhs) to be a TileType, but got "
                   << args[2]->GetType()->TypeName();

    // Extract shapes
    const auto &accShape = accType->shape_;
    const auto &lhsShape = lhsType->shape_;
    const auto &rhsShape = rhsType->shape_;

    // For block matmul_acc, we require 2D tiles
    CHECK(accShape.size() == 2) << "The operator " << opName << " requires acc to be 2D, but got " << accShape.size()
                                << " dimensions";
    CHECK(lhsShape.size() == 2) << "The operator " << opName << " requires lhs to be 2D, but got " << lhsShape.size()
                                << " dimensions";
    CHECK(rhsShape.size() == 2) << "The operator " << opName << " requires rhs to be 2D, but got " << rhsShape.size()
                                << " dimensions";

    // Matrix multiplication with accumulation: acc[M, N] += lhs[M, K] @ rhs[K, N]
    ExprPtr mDimAcc = accShape[0];
    ExprPtr nDimAcc = accShape[1];

    // Verify dimensions match
    auto mAccConst = As<ConstInt>(mDimAcc);
    auto mLhsConst = As<ConstInt>(lhsShape[0]);
    auto nAccConst = As<ConstInt>(nDimAcc);
    auto nRhsConst = As<ConstInt>(rhsShape[1]);
    auto kLhsConst = As<ConstInt>(lhsShape[1]);
    auto kRhsConst = As<ConstInt>(rhsShape[0]);

    if (mAccConst && mLhsConst) {
        CHECK(mAccConst->value_ == mLhsConst->value_)
            << "The operator " << opName << " requires matching M dimensions, but got acc M=" << mAccConst->value_
            << " and lhs M=" << mLhsConst->value_;
    }

    if (nAccConst && nRhsConst) {
        CHECK(nAccConst->value_ == nRhsConst->value_)
            << "The operator " << opName << " requires matching N dimensions, but got acc N=" << nAccConst->value_
            << " and rhs N=" << nRhsConst->value_;
    }

    if (kLhsConst && kRhsConst) {
        CHECK(kLhsConst->value_ == kRhsConst->value_)
            << "The operator " << opName << " requires matching K dimensions, but got lhs K=" << kLhsConst->value_
            << " and rhs K=" << kRhsConst->value_;
    }

    // Promote data types
    auto lhsRhsDtype = PromoteDataTypes(lhsType->dtype_, rhsType->dtype_);
    CHECK(lhsRhsDtype) << "The operator " << opName << " requires compatible lhs and rhs data types, but got "
                       << lhsType->dtype_.ToString() << " and " << rhsType->dtype_.ToString();

    auto resultDtype = PromoteDataTypes(accType->dtype_, *lhsRhsDtype);
    CHECK(resultDtype) << "The operator " << opName << " requires compatible accumulator data type, but got "
                       << accType->dtype_.ToString() << " and " << lhsRhsDtype->ToString();

    // Output shape is [M, N] (same as accumulator)
    std::vector<ExprPtr> outputShape = {mDimAcc, nDimAcc};

    return std::make_shared<TileType>(outputShape, *resultDtype);
}

// ============================================================================
// Registration Function for Block Matrix Multiplication Operations
// ============================================================================

REGISTER_OP("block.matmul")
    .SetOpCategory("BlockOp")
    .SetDescription("Matrix multiplication of two tiles")
    .AddArgument("lhs", "Left-hand side tile (TileType, 2D)")
    .AddArgument("rhs", "Right-hand side tile (TileType, 2D)")
    .SetDeduceType([](const std::vector<ExprPtr> &args, const std::vector<std::pair<std::string, std::any>> &kwargs) {
        return DeduceBlockMatMulType(args, kwargs, "block.matmul");
    });

REGISTER_OP("block.matmul_acc")
    .SetOpCategory("BlockOp")
    .SetDescription("Matrix multiplication with accumulation: acc = acc + lhs @ rhs")
    .AddArgument("acc", "Accumulator tile (TileType, 2D)")
    .AddArgument("lhs", "Left-hand side tile (TileType, 2D)")
    .AddArgument("rhs", "Right-hand side tile (TileType, 2D)")
    .SetDeduceType([](const std::vector<ExprPtr> &args, const std::vector<std::pair<std::string, std::any>> &kwargs) {
        return DeduceBlockMatMulAccType(args, kwargs, "block.matmul_acc");
    });

} // namespace ir
} // namespace pypto
